### PR TITLE
Add records_processed Prom metrics

### PR DESCRIPTION
**Pull request by busma13**
This PR makes the following changes:
- Adds `recordsWritten` variable and `getRecordsWritten` method to the `ElasticsearchBulk` processor
- Adds `elasticsearch_records_written` promMetric to the `ElasticsearchBulk` processor

- Adds `recordsFetched` variable and `getRecordsFetched` method to the `ElasticsearchReaderAPI`
- Adds `elasticsearch_records_read` promMetric to the `ReaderAPIFetcher`

### CODE REVIEW

File: asset/src/__lib/ReaderAPIFetcher.ts [hunk 1]
```diff
@@ -12,6 +12,20 @@ export class ReaderAPIFetcher extends Fetcher<ESDateConfig> {
         const apiManager = this.getAPI<ElasticReaderFactoryAPI>(apiName);
         this.api = await apiManager.create(apiName, {});
         await super.initialize();
+
+        const { context, api, opConfig } = this;
+        await this.context.apis.foundation.promMetrics.addGauge(
+            'records_processed',
+            'Number of records fetched',
+            ['operation'],
```
Comment (busma13) [inline asset/src/__lib/ReaderAPIFetcher.ts:20]:
```suggestion
            ['op_name'],
```

**Comment by busma13** [comment]
elasticsearch_bulk job:
```
{
    "name": "datagen-metrics-es",
    "lifecycle": "persistent",
    "workers": 1,
    "assets": [
        "standard",
        "elasticsearch:3.6.0"
    ],
    "prom_metrics_enabled": true,
    "prom_metrics_port": 3333,
    "prom_metrics_add_default": false,
    "operations": [
        {
            "_op": "data_generator",
            "size": 10000
        },
        {
            "_op": "elasticsearch_bulk",
            "index": "peter-test-metrics-local",
            "size": 10000
        }
    ],
    "__metadata": {
        "cli": {
            "cluster": "http://localhost:5678",
            "version": "1.2.0",
            "job_id": "6d110d40-8ddb-4797-bdb8-9ca62f58aff1",
            "updated": "2024-05-20T22:24:17.046Z"
        }
    }
}
```

scrape results:
```
ts-wkr-datagen-metrics-es-6d110d40-8ddb-6cd6948c4b-t6dzg:/app/source# curl localhost:3333/metrics
# HELP teraslice_worker_info Information about Teraslice worker
# TYPE teraslice_worker_info gauge
teraslice_worker_info{arch="arm64",clustering_type="kubernetes",name="teraslice",node_version="v18.19.1",platform="linux",teraslice_version="1.6.1",assignment="worker",ex_id="298834e0-8880-4440-b3da-043dfe4b90b5",job_id="6d110d40-8ddb-4797-bdb8-9ca62f58aff1",job_name="datagen-metrics-es",pod_name="ts-wkr-datagen-metrics-es-6d110d40-8ddb-6cd6948c4b-t6dzg"} 1

# HELP teraslice_worker_slices_processed Number of slices the worker has processed
# TYPE teraslice_worker_slices_processed gauge
teraslice_worker_slices_processed{name="teraslice",assignment="worker",ex_id="298834e0-8880-4440-b3da-043dfe4b90b5",job_id="6d110d40-8ddb-4797-bdb8-9ca62f58aff1",job_name="datagen-metrics-es",pod_name="ts-wkr-datagen-metrics-es-6d110d40-8ddb-6cd6948c4b-t6dzg"} 50

# HELP teraslice_worker_elasticsearch_records_written Number of records written to elasticsearch
# TYPE teraslice_worker_elasticsearch_records_written gauge
teraslice_worker_elasticsearch_records_written{op_name="elasticsearch_bulk",name="teraslice",assignment="worker",ex_id="298834e0-8880-4440-b3da-043dfe4b90b5",job_id="6d110d40-8ddb-4797-bdb8-9ca62f58aff1",job_name="datagen-metrics-es",pod_name="ts-wkr-datagen-metrics-es-6d110d40-8ddb-6cd6948c4b-t6dzg"} 500000
```

**Comment by busma13** [comment]
elasticsearch_reader job:
```
{
    "name": "metrics-es-to-kafka",
    "lifecycle": "once",
    "workers": 1,
    "assets": [
        "elasticsearch:3.6.0",
        "kafka"
    ],
    "prom_metrics_enabled": true,
    "prom_metrics_port": 3333,
    "prom_metrics_add_default": false,
    "operations": [
        {
            "_op": "elasticsearch_reader",
            "index": "peter-test-metrics-local",
            "date_field_name": "created",
            "query": "bytes:>=30000000",
            "start": "2010-08-04T10:00:00.000Z",
            "end": "2024-05-21T21:46:49.520Z"
        },
        {
            "_op": "delay",
            "ms": 1000
        },
        {
            "_op": "kafka_sender",
            "topic": "test_prom_metrics-v3",
            "size": 5000
        }
    ],
    "__metadata": {
        "cli": {
            "cluster": "http://localhost:5678",
            "version": "1.2.0",
            "job_id": "81d03e25-76bb-44e3-aa49-0dd260285f82",
            "updated": "2024-05-20T22:24:36.040Z"
        }
    }
}
```

scrape results:
```
ts-wkr-metrics-es-to-kafka-81d03e25-76bb-6fc985d699-g25nn:/app/source# curl localhost:3333/metrics
# HELP teraslice_worker_info Information about Teraslice worker
# TYPE teraslice_worker_info gauge
teraslice_worker_info{arch="arm64",clustering_type="kubernetes",name="teraslice",node_version="v18.19.1",platform="linux",teraslice_version="1.6.1",assignment="worker",ex_id="48efba80-d157-42ea-9dd6-1a6b6813fa6f",job_id="81d03e25-76bb-44e3-aa49-0dd260285f82",job_name="metrics-es-to-kafka",pod_name="ts-wkr-metrics-es-to-kafka-81d03e25-76bb-6fc985d699-g25nn"} 1

# HELP teraslice_worker_slices_processed Number of slices the worker has processed
# TYPE teraslice_worker_slices_processed gauge
teraslice_worker_slices_processed{name="teraslice",assignment="worker",ex_id="48efba80-d157-42ea-9dd6-1a6b6813fa6f",job_id="81d03e25-76bb-44e3-aa49-0dd260285f82",job_name="metrics-es-to-kafka",pod_name="ts-wkr-metrics-es-to-kafka-81d03e25-76bb-6fc985d699-g25nn"} 37

# HELP teraslice_worker_elasticsearch_records_read Number of records read from elasticsearch
# TYPE teraslice_worker_elasticsearch_records_read gauge
teraslice_worker_elasticsearch_records_read{op_name="elasticsearch_reader",name="teraslice",assignment="worker",ex_id="48efba80-d157-42ea-9dd6-1a6b6813fa6f",job_id="81d03e25-76bb-44e3-aa49-0dd260285f82",job_name="metrics-es-to-kafka",pod_name="ts-wkr-metrics-es-to-kafka-81d03e25-76bb-6fc985d699-g25nn"} 132864
```

**Review by godber — APPROVED** [review]

**Comment by godber** [comment]
The description needs updating to match the recent changes to the code/metric names.  Then I can merge.

**Pull request merged**